lib/ts/recipe/session/utils.ts

```diff
@@ -127,30 +127,58 @@ export function validateAndNormaliseUserInput(
     appInfo: NormalisedAppinfo,
     config?: TypeInput
 ): TypeNormalisedInput {
-    let cookieDomain =
-        config === undefined || config.cookieDomain === undefined
-            ? undefined
-            : normaliseSessionScopeOrThrowError(config.cookieDomain);
+    let cookieDomain = async (req: BaseRequest, userContext: any) => {
+        if (config !== undefined && config.cookieDomain !== undefined) {
+            if (typeof config.cookieDomain === "string") {
+                return normaliseSessionScopeOrThrowError(config.cookieDomain);
+            } else {
+                const domain = await config.cookieDomain(req, userContext);
+                return normaliseSessionScopeOrThrowError(domain);
+            }
+        }
+        return undefined;
+    };
     let accessTokenPath =
         config === undefined || config.accessTokenPath === undefined
             ? new NormalisedURLPath("/")
             : new NormalisedURLPath(config.accessTokenPath);
-    let protocolOfAPIDomain = getURLProtocol(appInfo.apiDomain.getAsStringDangerous());
-    let protocolOfWebsiteDomain = getURLProtocol(appInfo.websiteDomain.getAsStringDangerous());
-
-    let cookieSameSite: "strict" | "lax" | "none" =
-        appInfo.topLevelAPIDomain !== appInfo.topLevelWebsiteDomain || protocolOfAPIDomain !== protocolOfWebsiteDomain
-            ? "none"
-            : "lax";
-    cookieSameSite =
-        config === undefined || config.cookieSameSite === undefined
-            ? cookieSameSite
-            : normaliseSameSiteOrThrowError(config.cookieSameSite);
-
-    let cookieSecure =
-        config === undefined || config.cookieSecure === undefined
+    const cookieSameSite = async (req: BaseRequest, userContext: any) => {
+        const origin = await appInfo.origin(req, userContext);
+        const originString = origin.getAsStringDangerous();
+        let protocolOfAPIDomain = getURLProtocol(appInfo.apiDomain.getAsStringDangerous());
+        let protocolOfWebsiteDomain = getURLProtocol(originString);
+        let topLevelWebsiteDomain = getTopLevelDomainForSameSiteResolution(originString);
+        let cookieSameSiteNormalise: "strict" | "lax" | "none" = "none";
```

Comment:
Extract `"strict" | "lax" | "none"` into its own type if it doesnt already exist, Make sure `"strict" | "lax" | "none"` is not used directlly anywhere in code